### ISSUE TITLE
Abstract to remain resilient to failures

### ISSUE THREAD

**Issue by StarpTech** [apparatus/mu#65]
Mu is an message based router for building distributed systems this include a concept which describes how to dealing with failure scenerios.

Netflix use this http://techblog.netflix.com/2012/02/fault-tolerance-in-high-volume.html I think we can use it to implement a simliar strategy for Mu.

-  Request queue
-  [Circuit breaker](https://github.com/Netflix/Hystrix/wiki/How-it-Works#circuit-breaker)
-  [Calculate Circuit Health](https://github.com/Netflix/Hystrix/wiki/How-it-Works#7-calculate-circuit-health)
- Fallback solutions

Now, we can start the discussion.

References:
https://github.com/Netflix/Hystrix/wiki/Operations
http://techblog.netflix.com/2012/02/fault-tolerance-in-high-volume.html
https://msdn.microsoft.com/en-us/library/dn589784.aspx
https://github.com/Netflix/Hystrix/wiki/

**Comment by davidmarkclements** [comment]
Yep absolutely right, these things have been on the radar for some time, 
great time to start outlining an approach cc @mcollina

**Comment by StarpTech** [comment]
Related https://github.com/apparatus/mu/issues/55

**Comment by StarpTech** [comment]
Good circuit breaker nodejs implementation
http://yammer.github.io/circuit-breaker-js/

**Comment by mcollina** [comment]
@StarpTech that was good research! In the mu terminology, where do you see this?
I think it is in transports, either as a wrapping transport `mu.outbound({ hello: 'world' }, circuitBreaker(mytransport))`. Or it can live in the `mu` core. I am more on the wrapping transport side.

As the implementation goes, I would prefer to have our own implementation. Would you like to contribute?

**Comment by davidmarkclements** [comment]
I think it would be an adapter, like mu-balance or mu-tee

**Comment by davidmarkclements** [comment]
@StarpTech if you would like to build, I'll add you to contribs, a circuit breaker def needs to be part of the core monorepo

**Comment by StarpTech** [comment]
@mcollina I need some investigation to understand the whole concept of Mu but its very simliar to seneca so I think its blink. I also prefer to build it as own package. @davidmarkclements  Yes, would be great to be able to contribute. At the moment Im bussy but I will commit regular intervals like here.

**Comment by mcollina** [comment]
I think this concept should be in core!

Dave, we might want to setup an empty package folder here now, and open an issue to track progess.

**Comment by davidmarkclements** [comment]
yah circuit breaker is def core - will do tomorrow

**Comment by pelger** [comment]
Agreed that these are required @StarpTech. These should be implemented as transport adapters, and should live in the core mono-repo - as per the example balance and tee adapters.
